src/ddsrt/src/process/windows/process.c

```diff
@@ -0,0 +1,374 @@
+                                 &dupHandle,
+                                 PROCESS_ALL_ACCESS,
+                                 FALSE,
+                                 0);
+    if (duplicated) {
+      phdl = dupHandle;
+    }
+
+    rv = process_wait_exit(phdl, 0, NULL);
+    if (rv == DDS_RETCODE_TIMEOUT) {
+      /* Target process is still alive. Try to terminate it by
+       * injecting a thread that will execute a ExitProcess(). */
+      FARPROC funcExitProcess;
+      funcExitProcess = GetProcAddress(kernel32, "ExitProcess");
+      if (funcExitProcess) {
+        remoteHandle = CreateRemoteThread(phdl,
+                                          NULL,
+                                          0,
+                                          (LPTHREAD_START_ROUTINE)funcExitProcess,
+                                          (LPVOID)1,  /* Exit code */
+                                          0,
+                                          &threadId);
+        if (remoteHandle == 0) {
+          rv = DDS_RETCODE_ERROR;
+        } else {
+          /* Wait for the process to exit. */
+          waitmsec = INFINITE;
+          if (timeout != DDS_INFINITY) {
+            waitmsec = (DWORD)(timeout / DDS_NSECS_IN_MSEC);
+          }
+          waitres = WaitForSingleObject(phdl, waitmsec);
+          if (waitres == 0) {
+            rv = DDS_RETCODE_OK;
+          } else if (waitres == WAIT_TIMEOUT) {
+            rv = DDS_RETCODE_TIMEOUT;
+          } else {
+            rv = DDS_RETCODE_ERROR;
+          }
+          CloseHandle(remoteHandle);
+        }
+      } else {
+        rv = DDS_RETCODE_ERROR;
+      }
+    }
+
+    if (duplicated) {
+      CloseHandle(dupHandle);
+    }
+  }
+
+  return rv;
+}
+
+
+
+/* Forcefully kill the given process. */
+static dds_retcode_t
+process_kill(
+  HANDLE phdl,
+  dds_duration_t timeout)
+{
+  assert(phdl != 0);
+  if (TerminateProcess(phdl, 1 /* exit code */) == 0) {
+    return DDS_RETCODE_OK;
+  }
+  return DDS_RETCODE_ERROR;
+}
+
+
+
+/* Add quotes to a given string. */
+static size_t
+stringify_len(const char *str)
+{
+  return strlen(str) + 4; /* '\"' + str + '\"' + ' ' + '\0' */
+}
+static char*
+stringify(char *buf, const char *str)
+{
+  size_t len = strlen(str);
+  size_t idx = 0;
+  buf[idx++] = '\"';
+  if (len > 0) {
+    memcpy(&(buf[idx]), str, len);
+    idx += len;
+  }
+  buf[idx++] = '\"';
+  buf[idx++] = ' ';
+  buf[idx  ] = '\0';
+  return &(buf[idx]);
+}
+
+
+
+/* Create command line with executable and arguments. */
+static char*
+commandline(const char *exe, char *const argv_in[])
+{
+  char *cmd;
+  size_t cmdlen;
+
+  assert(exe);
+  cmdlen = stringify_len(exe);
+
+  if (argv_in == NULL) {
+    cmd = ddsrt_malloc(cmdlen);
+    stringify(cmd, exe);
+  } else {
+    char *ptr;
+    size_t argi;
+    for (argi = 0; argv_in[argi] != NULL; argi++) {
+      cmdlen += stringify_len(argv_in[argi]);
+    }
+    cmd = ddsrt_malloc(cmdlen);
+    ptr = stringify(cmd, exe);
```

Comment:
It'd be interesting to test what happens if I pass an argument that contains a double-quote character. Did you test?

Reply:
An argument with a backslash character could be funny also...

Reply:
You have a weird sense of 'funny' :)

Reply:
I did a quick check and it went as you might expect. On Posix, it works fine. On Windows, not so much. This'll need fixing.

Reply:
Tests are added and issue has been fixed.